airflow/bin/cli.py

```
@@ -328,19 +328,67 @@ def run(args, dag=None):
+
+        # To handle log writing when tasks are impersonated, the log files need to
+        # be writable by the user that runs the Airflow command and the user
+        # that is impersonated. This is mainly to handle corner cases with the
+        # SubDagOperator. When the SubDagOperator is run, all of the operators
+        # run under the impersonated user and create appropriate log files
+        # as the impersonated user. However, if the user manually runs tasks
+        # of the SubDagOperator through the UI, then the log files are created
+        # by the user that runs the Airflow command. For example, the Airflow
+        # run command may be run by the `airflow_sudoable` user, but the Airflow
+        # tasks may be run by the `airflow` user. If the log files are not
+        # writable by both users, then it's possible that re-running a task
+        # via the UI (or vice versa) results in a permission error as the task
+        # tries to write to a log file created by the other user.
+        log_base = os.path.expanduser(conf.get('core', 'BASE_LOG_FOLDER'))
+        directory = log_base + "/{args.dag_id}/{args.task_id}".format(args=args)
+        # Create the log file and give it group writable permissions
+        # TODO(aoen): Make log dirs and logs globally readable for now since the SubDag
+        # operator is not compatible with impersonation (e.g. if a Celery executor is used
+        # for a SubDag operator and the SubDag operator has a different owner than the
+        # parent DAG)
+        if not os.path.exists(directory):
+            # Create the directory as group writable using custom mkdirs
+            # as os.makedirs doesn't set mode properly.
+            mode = 0x775
+            mkdirs(directory, 0o777)
+        iso = args.execution_date.isoformat()
+        filename = "{directory}/{iso}".format(**locals())
+
+        if not os.path.exists(filename):
+            open(filename, "a").close()
+            mode = 0o660
```

To cover all possible cases with the SubDagOperator, I think the log file needs to be globally readable and writable. One example case is if the user runs a task as one user, then later decides to run the task as another user for the same execution date.

---

Done, interestingly this isn't the case in our prod branch at airbnb (it's 664 which is not world-writeable).

---

That should be relatively simple to solve by verifying and changing the owner or group if needed.

---

Sorry, I should have elaborated more on the potential case with the SubDagOperator. If you use the SubDagOperator and the CeleryExecutor, the task runs as the `run_as_user` specified for the task. Because that's not a privileged user, it can't change the permission for the log file.

---

My suggestion would then be to not support this for the 1) subdagoperator (throw an exception so the executor can log it), 2) let the executor (worker) do the change which can be privileged, or 3) use a logging facility like syslog. 

1 seems simplest until the subdagoperator is changed, 2 would be a kind of workaround and 3 would be a kind of overhaul (which is needed in the logging area in general).

---

I think too many users rely on the operator to throw an exception instead of support logging unfortunately. Let's start by supporting full compatibility with existing infrastructure and then continue to tighten the looseness in security (e.g. this and supporting moving airflow secrets to a vault).